tracer/src/Datadog.Trace/Processors/NormalizerTraceProcessor.cs

```diff
@@ -52,9 +52,12 @@ public ArraySegment<Span> Process(ArraySegment<Span> trace)
                 +----------+--------------------------------------------------------------------------------------------------------+
              */
 
-            for (var i = trace.Offset; i < trace.Count + trace.Offset; i++)
+            // https://github.com/DataDog/datadog-agent/blob/eac2327c5574da7f225f9ef0f89eaeb05ed10382/pkg/trace/agent/normalizer.go#L133-L135
+            var traceContext = trace.Array![trace.Offset].Context.TraceContext;
+
+            if (!string.IsNullOrEmpty(traceContext.Environment))
             {
-                trace.Array![i] = Process(trace.Array[i]);
+                traceContext.Environment = TraceUtil.NormalizeTag(traceContext.Environment);
```

Comment:
Nitpicking but do you think we should cache the normalized env as well?

Reply:
Definitely, thank you.

Reply:
We now truncate and normalize the tag value before caching the encoded bytes.

Reply:
🤔 , I had in mind to do the memoization here in `NormalizeTraceProcessor`. Because we're likely to have only a few environments used by the customer. But we don't apply this normalization on every traces. If what I see from the code is true, we do the normalization only in CI Processors and when Stats are enabled.

So moving it to the SpanMessagePackFormatter brings 2 issues I believe. One it's modifying the current behaviour to normalize everything. Shouldn't be a big deal as the trace agent does the same but it means we're doing extra work that was handled by the agent. 
Also, it means that we add functional logic (normalization) in a wiring only one (MessagePackFormatter). Sorry for the back and forth, I should have made my first comment clearer.

Reply:
I thought about this before making the change and...

> we do the normalization only in CI Processors and when Stats are enabled... it's modifying the current behaviour to normalize everything. Shouldn't be a big deal as the trace agent does the same but it means we're doing extra work that was handled by the agent.

True, but the result is cached so the extra work is amortized over time. Ideally, it only happens once for the lifetime of the service.

> we add functional logic (normalization) in a wiring only one (MessagePackFormatter)

Also true, and the normalization is needed for stats, too, so doing it only for serialization is too late. I am reverting the normalization change for now and we can tackle it in a separate PR.